[Arvind Natarajan] iP

DiegoTheExplorar opened the pull request:
## iP Week 4 🚀

> "Am I doing this right?" - DiegoTheExplorar

### Changes Made

- Updated the `Launcher` class
- Implemented the `MainApp` class

### Next Steps

1. Finish Week 5 iP


### Code Example

```java
public class Launcher {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

### Tasks

- [x] Finish iP tasks till **Week 4**
- [ ] Finish iP tasks till *Week 5*


~~I didn't know where to add a strikethrough~~

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

DiegoTheExplorar committed:
Complete Week 2 Level-0 tasks

Implemented basic chatbot functionality by adding greeting and exit messages.
Renamed the chatbot class to reflect the new name and removed all instances
of 'Duke' from the codebase.

DiegoTheExplorar committed:
Finish Week 2 Level-1

Added in the echo feature

DiegoTheExplorar committed:
Finished Week 2 Level 2: Add, List

DiegoTheExplorar committed:
Finish Week 2 Level 3

Added a Task class and Mark as Done

DiegoTheExplorar committed:
Finish Level 4

Added in new classes like todo, event and deadline

DiegoTheExplorar committed:
Finish A-TextUiTesting: Automated Text UI Testing

DiegoTheExplorar committed:
Finish Level-5: Handle Errors

Added in classes like NoDescriptionException and UnknownCommandException to handle errors

DiegoTheExplorar committed:
Finish Level-6: Delete

Changed data structure to ArrayList and implemented Delete

DiegoTheExplorar committed:
Add file overwrite functionality and scanner closure

Implemented automatic file overwrite after each task update to ensure data persistence. Added proper closure of scanner after reading tasks from the file to prevent resource leaks.

DiegoTheExplorar committed:
Support multiple datetime formats with optional time

Added support for parsing deadlines with various date and time
formats. If a time is provided in the input, it is stored as a
LocalDateTime. If no time is provided, only the date is stored as a
LocalDate. Updated toString() method to format deadlines with or
without time based on input. Ensured that deadlines are correctly
saved and loaded from file with the appropriate format.

DiegoTheExplorar committed:
Improved OOP Design

Extracted Ui class to handle user interactions.
Created Storage class to manage task file operations. Introduced Parser class to interpret user commands. Added TaskList class for managing tasks, including add/delete operations.

DiegoTheExplorar committed:
Finish A-Packages

Put all the classes in diego folder

DiegoTheExplorar committed:
Merge branch 'add-gradle-support'

DiegoTheExplorar committed:
Finish A-JUnit

Set up of gradle and added 2 JUnit test files

DiegoTheExplorar committed:
Add JavaDoc comments to non-private classes and methods

Added JavaDoc comments to all non-private classes and methods across the codebase. Included descriptions for constructors, methods, and fields to improve code readability and maintainability. Documented the purpose of each class and its main responsibilities. Provided detailed explanations for non-trivial private methods. Ensured consistency in the format and style of the JavaDoc comments.

This commit enhances the overall documentation quality, making it easier for developers to understand and work with the code.

DiegoTheExplorar committed:
Add "find" command to search for tasks by keyword

- Implemented `FindCommand` class to allow users to search for tasks containing a specific keyword.
- Updated `Parser` to recognize the `find` command.
- Modified `TaskList` to include a `findTasks` method that returns matching tasks.
- Enhanced `Ui` with a method to display the results of the search.

This update improves the functionality of the application by enabling keyword-based task searches.

DiegoTheExplorar committed:
Fix error in FindCommand.java

Command's execute return has no return value hence FindCommand's should return the same

DiegoTheExplorar committed:
Merge remote-tracking branch 'origin/branch-Level-9'

DiegoTheExplorar committed:
I didnt resolve conflicts properly

DiegoTheExplorar committed:
Finish A-CheckStyle

Added xml files and tested that it works

DiegoTheExplorar committed:
Merge branch 'A-CheckStyle'

* A-CheckStyle:
  Finish A-CheckStyle

DiegoTheExplorar committed:
Finish JFX GUI

- Refactored all classes for clarity
- Changed command and ui classes to return string
- Added in JFX

DiegoTheExplorar committed:
Revert "Merge branch 'A-CheckStyle'"

This reverts commit 1e310474badfad5009538f96274b8d16546d39a3, reversing
changes made to 813320c08ccb52df7c9492f39429fcbd1cf9eded.

DiegoTheExplorar committed:
Merge branch 'Level-10'

ZweZeya commented on src/main/java/Tasks/Deadline.java:
```diff
@@ -0,0 +1,102 @@
+            DateTimeFormatter.ofPattern("MMM d yyyy")
+    );
+
+    protected static final DateTimeFormatter outputFormatterWithTime = DateTimeFormatter.ofPattern("MMM dd yyyy, h:mm a");
```
I think you can bring this to the next line.

ZweZeya commented on src/main/java/Commands/AddTodoCommand.java:
```diff
@@ -0,0 +1,38 @@
+import Tasks.Task;
+import Tasks.TaskList;
+import Tasks.Todo;
+import Main.*;
```
Try not to use wildcard imports.

ZweZeya reviewed (commented):
The code is generally well-done. I did notice quite a some wildcard imports, including the one I pointed out, so you can take a look at that. For the formatting, you can try to break down longer lines as per the style guide. Javadocs are clear and concise!

DiegoTheExplorar committed:
Add assertions to TaskList, Storage, Deigo, MainWindow, and DialogBox
classes

- Added assertions in TaskList to ensure correct task handling and data
integrity.
- Included assertions in Storage to validate data storage and retrieval
processes.
- Implemented assertions in Deigo to verify critical operations and
state changes.
- Introduced assertions in MainWindow to check UI state and interactions.
- Added assertions in DialogBox to validate user input and dialog
behavior.

These updates improve the reliability and robustness of the application
by providing early detection of issues and ensuring that all critical
components perform as expected. Assertions serve as a safety net
during development and debugging, helping to catch potential problems
before they impact end-users.

DiegoTheExplorar committed:
Merge pull request #1 from DiegoTheExplorar/heads/A-Assertions

Add assertions to TaskList, Storage, Deigo, MainWindow, and DialogBox classes

DiegoTheExplorar committed:
Improve code quality and style in Deadline and Storage classes

This commit refactors the Deadline and Storage classes to enhance code
quality and adherence to style guidelines.

- **Reordered Import Statements:** Organized import statements in
  alphabetical order and separated standard Java imports from third-party
  and custom imports for clarity.

- **Updated Static Variables:** Reordered static final variables in the
  Deadline class to follow consistent naming and ordering conventions:
  moved `DATE_FORMATS_WITH_TIME` and `DATE_FORMATS_WITHOUT_TIME` above
  output formatters.

- **Fixed Formatting Issues:**
  - Ensured all lines adhere to a maximum length of 120 characters.
  - Corrected indentation and removed trailing whitespace to align
    with project style guidelines.

- **Enhanced Code Style:**
  - Added missing default clauses in switch statements to handle
    unexpected cases.
  - Updated naming conventions for constants to use uppercase with
    underscores.

- **Refactored Import Statements in Storage Class:** Organized imports to
  follow proper ordering and avoided the use of wildcard imports.

DiegoTheExplorar committed:
Merge pull request #2 from DiegoTheExplorar/branch-CodeQuality

Improve code quality and style in Deadline and Storage classes

DiegoTheExplorar committed:
Implement remind feature for tasks due within a specified number of days

Added support for the 'remind' command, allowing users to receive
reminders for tasks that are due within a given number of days.
The default behavior reminds users of tasks due within 7 days
if no specific number is provided.

- Introduced a new ReminderCommand class that accepts a dynamic
number of days.
- Updated Task and Deadline classes to correctly handle date
comparisons for upcoming tasks.
- Modified the Parser to handle 'remind [days]' commands,
defaulting to 7 days if not specified.
- Updated the Ui class to display tasks due within the
specified timeframe.

DiegoTheExplorar committed:
Merge pull request #3 from DiegoTheExplorar/branch-A-CI

Implement remind feature for tasks due within a specified number of days

choaticman commented on src/main/java/Commands/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,41 @@
+package Commands;
+
+import Main.Storage;
+import Main.Ui;
+import Tasks.Deadline;
+import Tasks.Task;
+import Tasks.TaskList;
```
Perhaps the name of packages should be in all lowercase, to avoid confusion between Packages and Classes.

choaticman commented on src/main/java/Commands/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,41 @@
+/**
+ * Represents a command to add a deadline task to the task list.
+ */
+public class AddDeadlineCommand implements Command {
```
Perhaps the name of a class should be a noun. (E.g. DeadlineCommandAdder)

choaticman commented on src/main/java/Main/Parser.java:
```diff
@@ -0,0 +1,89 @@
+     * @return The appropriate Command object based on the user input.
+     * @throws DiegoException If the input is invalid or unknown.
+     */
+    public Command parseCommand(String input) throws DiegoException {
```
Perhaps you could use a switch-case table to improve readability.

choaticman commented on src/main/java/Main/Parser.java:
```diff
@@ -0,0 +1,89 @@
+    private int parseIndex(String input) throws DiegoException {
+        try {
+            return Integer.parseInt(input.split(" ")[1]) - 1;
+        } catch (Exception e) {
```
Should there be a more specific `Exception` to catch to avoid Pokemon Exception Handling?

choaticman reviewed (commented):
Overall the code as well as most of the naming of methods and classes LGTM. Some room for improvements would include naming of packages in all lowercase, as well as using nouns for Classes' names.

DiegoTheExplorar committed:
Refactor code for improved readability and maintainability

Refactored multiple command classes to adhere to good coding practices.
Changes include:
- Improved variable naming for better clarity.
- Added meaningful comments where necessary and removed redundant ones.
- Ensured immutability by marking fields as `final` where appropriate.
- Extracted complex logic into smaller methods to follow the
Single Responsibility Principle.
- Enhanced error handling with more descriptive exceptions.

This refactor improves the overall maintainability and clarity of the
code, following the SE-EDU Java coding standards.

DiegoTheExplorar committed:
Fix ReminderCommand bugs and ShadowJar debug

- There was an issue regarding the wrong 'days' printed when
remind command was used

DiegoTheExplorar committed:
Set up .yml file for GitHub Actions CI/CD

DiegoTheExplorar committed:
Revert "Set up .yml file for GitHub Actions CI/CD"

This reverts commit 31297ff6765ead2ce68400a415844cd1084964a9.

DiegoTheExplorar committed:
Finish A-BetterGui

Added a HelpCommand class and made very minor tweaks to GUI

DiegoTheExplorar committed:
Finish A-MoreTesting

Added 2 JUnit tests. 1 for RemindCommand and 1 for HelpCommand

DiegoTheExplorar committed:
Merge branch 'branch-A-MoreTesting'

DiegoTheExplorar committed:
Finish A-UserGuide

Added the following:
- Ui picture
- Updated the README with user guide
